src/StableBuffer.mo

```diff
@@ -86,88 +89,95 @@ module {
   };
 
   /// Returns the count of elements in the buffer
-  public func size<X>(buffer: StableBuffer<X>) : Nat { buffer.count };
+  public func size<X>(buffer : StableBuffer<X>) : Nat { buffer.count };
 
   /// Resets the buffer.
-  public func clear<X>(buffer: StableBuffer<X>): () {
+  public func clear<X>(buffer : StableBuffer<X>) : () {
     buffer.count := 0;
   };
 
   /// Returns a copy of this buffer.
-  public func clone<X>(buffer: StableBuffer<X>) : StableBuffer<X> {
+  public func clone<X>(buffer : StableBuffer<X>) : StableBuffer<X> {
     let c = initPresized<X>(buffer.elems.size());
-    var i = 0;
+    var i = vals(buffer);
+
     label l loop {
-      if (i >= buffer.count) break l;
-      add(c, buffer.elems[i]);
-      i += 1;
+      switch (i.next()) {
+        case null break l;
+        case (?x) { add(c, x) };
+      };
     };
-    c
+
+    c;
   };
 
   /// Returns an `Iter` over the elements of this buffer.
-  public func vals<X>(buffer: StableBuffer<X>) : { next : () -> ?X } = object {
+  public func vals<X>(buffer : StableBuffer<X>) : { next : () -> ?X } = object {
     var pos = 0;
     public func next() : ?X {
       if (pos == buffer.count) { null } else {
-        let elem = ?buffer.elems[pos];
+        let elem = buffer.elems[pos];
         pos += 1;
-        elem
-      }
-    }
+        elem;
+      };
+    };
   };
 
   /// Creates a Buffer from an Array
-  public func fromArray<X>(xs: [X]): StableBuffer<X> {
-    let ys: StableBuffer<X> = initPresized(xs.size());
+  public func fromArray<X>(xs : [X]) : StableBuffer<X> {
+    let ys : StableBuffer<X> = initPresized(xs.size());
     for (x in xs.vals()) {
       add(ys, x);
     };
 
-    ys
+    ys;
   };
 
   /// Creates a new array containing this buffer's elements.
-  public func toArray<X>(buffer: StableBuffer<X>) : [X] =
-    // immutable clone of array
-    Prim.Array_tabulate<X>(
-      buffer.count,
-      func(x : Nat) : X { buffer.elems[x] }
-    );
+  public func toArray<X>(buffer : StableBuffer<X>) : [X] =
+  // immutable clone of array
+  Prim.Array_tabulate<X>(
+    buffer.count,
+    func(x : Nat) : X {
+      get(buffer, x);
```

Comment:
All of these changes to use `get` add an extra function call (small amount of overhead). Not a huge deal, but directly using the `buffer.elems[index]` isn't that much cognitive overload.

Just thought I'd mention in case you have a strong opinion on this.

Reply:
The reason for using `get()` in this function is to remove the option type. We would still have to use the `switch` statement if we use `buffer.elems[index]` directly and duplicate the `Prim.trap` message.